File: LeetCodeSolutions/LeetCode_0112.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if not root:
            return False
        from collections import deque
        que = deque()
        que.append([root, root.val])
        while que:
            node, cur = que.popleft()
            if node.right:
                que.append([node.right, cur + node.right.val])
            if node.left:
                que.append([node.left, cur + node.left.val])
            if not node.right and not node.left:
                if cur == sum:
                    return True
        return False

File: LeetCodeSolutions/LeetCode_0112.cpp
/*
 * @lc app=leetcode.cn id=112 lang=cpp
 *
 * [112] 路径总和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (!root) return false;
        return helper(root, 0, sum);
    }

    bool helper(TreeNode *node, int cur_sum, int &sum){
        if(!node) return false;
        if (!node->left && !node->right) return cur_sum + node->val == sum;
        return helper(node->left, cur_sum + node->val, sum) || helper(node->right, cur_sum + node->val, sum);
    }
};
// @lc code=end

